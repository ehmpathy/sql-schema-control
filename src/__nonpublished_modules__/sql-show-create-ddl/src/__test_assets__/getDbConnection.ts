import { DatabaseLanguage } from '../../../../domain';
import { connectToDatabase } from './connectToDatabase';

// the connection config for our integration test dbs, provisioned by docker
export const config = {
  [DatabaseLanguage.MYSQL]: {
    host: 'localhost',
    port: 12821,
    database: 'superimportantdb',
    username: 'root',
    password: 'a-secure-password',
  },
  [DatabaseLanguage.POSTGRES]: {
    host: 'localhost',
    port: 7821,
    database: 'superimportantdb',
    schema: 'public',
    username: 'postgres',
    password: 'a-secure-password',
  },
};

export const getDbConnection = async ({
  language,
}: {
  language: DatabaseLanguage;
}) => connectToDatabase({ language, config: config[language] });
