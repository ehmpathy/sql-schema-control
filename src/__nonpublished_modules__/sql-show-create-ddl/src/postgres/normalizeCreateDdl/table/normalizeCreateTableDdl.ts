import sqlFormatter from 'sql-formatter';

/**
 * normalize generated create table ddl for comparisons
 *
 * TODO: name this function specific to its use case
 * e.g., "normalizeForSchemaComparison" or something similar
 */
export const normalizeCreateTableDdl = ({ ddl }: { ddl: string }) => {
  // prettify that ddl, to make standard spacing and appearance of ddl
  let prettierDdl = sqlFormatter.format(ddl);

  // swap back to better aliases
  prettierDdl = prettierDdl.replace(/int NOT NULL DEFAULT nextval\([\'\w: \n]+\)/g, 'serial NOT NULL'); // e.g., `bigint DEFAULT nextval('test_tb_for_show_create_on_id_seq' :: regclass)` -> `bigserial`; although 'bigint default ...' is the canonical def, its just too verbose to be useful
  prettierDdl = prettierDdl.replace(/character varying/g, 'varchar'); // although `character varying` is the canonical term, `varchar` is used more regularly in practice -> more helpful
  prettierDdl = prettierDdl.replace(/ integer /g, ' int '); // postgres converts `int` to `integer`, but `bigint` is kept as `bigint`.... lets be consistent and just use the shorthand
  prettierDdl = prettierDdl.replace(/ ?:: ?\w+( ?\[\])?/g, ''); // remove any ":: __TYPE__" casting that could exist in the DDL. this level of information has not been found as useful yet

  // postgres stores CHECK ($COLUMN_NAME IN ('A', 'B', 'C')) in an ugly way. change it back for ease of reading
  prettierDdl = prettierDdl.replace(
    /CHECK \([\s\n]+\([\s\n]+\(([\w_]+)\) = ANY \([\s\n]+\([\s\n]+ ARRAY \[([\w\d', -_]+)\][\s\n]+\)[\s\n]+\)[\s\n]+\)[\s\n]+\)/g,
    'CHECK ($1 IN ($2))',
  );

  // replace `REFERENCES table_name(id)` with `REFERENCES table_name (id)` for consistency
  prettierDdl = prettierDdl.replace(/REFERENCES ([\w_]+)\(([\w_]+)\)/g, 'REFERENCES $1 ($2)');

  // reprettify the ddl
  const reprettyDdl = sqlFormatter.format(prettierDdl);

  // return the normalized ddl
  return reprettyDdl;
};
