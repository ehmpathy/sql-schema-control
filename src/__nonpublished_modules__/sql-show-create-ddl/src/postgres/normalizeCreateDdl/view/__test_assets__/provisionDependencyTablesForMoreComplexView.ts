import { DatabaseConnection } from '../../../../types';

const dependencyTablesDdl = [
  `
CREATE TABLE home (
  id bigserial NOT NULL,
  uuid uuid NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  name varchar(255) NOT NULL,
  built timestamp with time zone NOT NULL,
  bedrooms int NOT NULL,
  bathrooms int NOT NULL,
  host_ids_hash bytea NOT NULL,
  CONSTRAINT home_pk PRIMARY KEY (id),
  CONSTRAINT home_ux1 UNIQUE (name, host_ids_hash)
);
      `.trim(),
  `
CREATE TABLE home_version (
  id bigserial NOT NULL,
  home_id bigint NOT NULL,
  effective_at timestamp with time zone NOT NULL DEFAULT now(),
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  photo_ids_hash bytea NOT NULL,
  CONSTRAINT home_version_pk PRIMARY KEY (id),
  CONSTRAINT home_version_ux1 UNIQUE (home_id, effective_at, created_at),
  CONSTRAINT home_version_fk0 FOREIGN KEY (home_id) REFERENCES home (id)
);
CREATE INDEX home_version_fk0_ix ON home_version USING btree (home_id);
      `.trim(),
  `
CREATE TABLE home_version_to_photo (
  id bigserial NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  home_version_id bigint NOT NULL,
  photo_id bigint NOT NULL,
  array_order_index smallint NOT NULL,
  CONSTRAINT home_version_to_photo_pk PRIMARY KEY (id),
  CONSTRAINT home_version_to_photo_ux1 UNIQUE (home_version_id, array_order_index),
  CONSTRAINT home_version_to_photo_fk0 FOREIGN KEY (home_version_id) REFERENCES home_version (id)
  -- CONSTRAINT home_version_to_photo_fk1 FOREIGN KEY (photo_id) REFERENCES photo (id)
);
CREATE INDEX home_version_to_photo_fk0_ix ON home_version_to_photo USING btree (home_version_id);
CREATE INDEX home_version_to_photo_fk1_ix ON home_version_to_photo USING btree (photo_id);
      `.trim(),
  `
CREATE TABLE home_to_host (
  id bigserial NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  home_id bigint NOT NULL,
  host_id bigint NOT NULL,
  array_order_index smallint NOT NULL,
  CONSTRAINT home_to_host_pk PRIMARY KEY (id),
  CONSTRAINT home_to_host_ux1 UNIQUE (home_id, array_order_index),
  CONSTRAINT home_to_host_fk0 FOREIGN KEY (home_id) REFERENCES home (id)
  -- CONSTRAINT home_to_host_fk1 FOREIGN KEY (host_id) REFERENCES host (id)
);
CREATE INDEX home_to_host_fk0_ix ON home_to_host USING btree (home_id);
CREATE INDEX home_to_host_fk1_ix ON home_to_host USING btree (host_id);
      `.trim(),
  `
CREATE TABLE home_cvp (
  id bigserial NOT NULL,
  updated_at timestamp with time zone NOT NULL DEFAULT now(),
  home_id bigint NOT NULL,
  home_version_id bigint NOT NULL,
  CONSTRAINT home_cvp_pk PRIMARY KEY (id),
  CONSTRAINT home_cvp_ux1 UNIQUE (home_id),
  CONSTRAINT home_cvp_fk0 FOREIGN KEY (home_id) REFERENCES home (id),
  CONSTRAINT home_cvp_fk1 FOREIGN KEY (home_version_id) REFERENCES home_version (id)
);
CREATE INDEX home_cvp_fk0_ix ON home_cvp USING btree (home_id);
CREATE INDEX home_cvp_fk1_ix ON home_cvp USING btree (home_version_id);
      `.trim(),
];
const dependencyTablesNames = [
  'home_cvp',
  'home_to_host',
  'home_version_to_photo',
  'home_version',
  'home',
];
export const provisionDependencyTablesForMoreComplexView = async ({
  dbConnection,
}: {
  dbConnection: DatabaseConnection;
}) => {
  await Promise.all(
    dependencyTablesNames.map((tableName) =>
      dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tableName} CASCADE` }),
    ),
  );
  await Promise.all(
    dependencyTablesDdl.map((ddl) => dbConnection.query({ sql: ddl })),
  );
};
