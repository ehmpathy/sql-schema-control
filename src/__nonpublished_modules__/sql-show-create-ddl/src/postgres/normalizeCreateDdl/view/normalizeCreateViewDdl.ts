import sqlFormatter from 'sql-formatter';
import { recursivelyHeavilyNormalizeViewDdl } from '../../../generic/recursivelyHeavilyNormalizeViewDdl/recursivelyHeavilyNormalizeViewDdl';

/**
 * normalize generated create table ddl for comparisons
 *
 * TODO: name this function specific to its use case
 * e.g., "normalizeForSchemaComparison" or something similar
 */
export const normalizeCreateViewDdl = ({ ddl }: { ddl: string }) => {
  let prettierDdl = ddl;

  // remove any ":: __TYPE__" casting that could exist in the DDL. this level of information has not been found as useful yet
  prettierDdl = prettierDdl.replace(/ ?:: ?\w+( ?\[\])?/g, '');

  // remove the extra parens that postgres's viewdef fn adds in
  prettierDdl = recursivelyHeavilyNormalizeViewDdl({ ddl: prettierDdl });

  // format that ddl, to make standard spacing and appearance of ddl
  prettierDdl = sqlFormatter.format(prettierDdl);

  // make sure that we dont have CREATE & OR REPLACE on two lines
  prettierDdl = prettierDdl.replace(/CREATE\nOR REPLACE/, 'CREATE OR REPLACE');

  // return the normalized ddl
  return prettierDdl;
};
