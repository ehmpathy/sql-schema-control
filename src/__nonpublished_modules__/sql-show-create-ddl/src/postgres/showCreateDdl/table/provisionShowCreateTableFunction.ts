import fs from 'fs';

import { DatabaseConnection } from '../../../types';

/**
 * fortunately for us, postgres does not come with built in `SHOW CREATE TABLE/VIEW/FUNCTION`
 *
 * we have to provide a way to do so manually. on the bright side, postgres does provide an extensive catalog system that we can leverage
 *
 * based on https://stackoverflow.com/questions/2593803/how-to-generate-the-create-table-sql-statement-for-an-existing-table-in-postgr
 */
export const provisionShowCreateTableFunction = async ({ dbConnection }: { dbConnection: DatabaseConnection }) => {
  const sql = await new Promise<string>((resolve, reject) =>
    fs.readFile(`${__dirname}/show_create_table.sql`, (error, data) => {
      if (error) return reject(error);
      return resolve(data.toString());
    }),
  );
  await dbConnection.query({ sql });
};
