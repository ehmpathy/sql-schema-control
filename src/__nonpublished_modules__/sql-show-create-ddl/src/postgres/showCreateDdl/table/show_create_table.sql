
CREATE OR REPLACE FUNCTION public.show_create_table(
  in_schema_name varchar,
  in_table_name varchar
)
RETURNS text
LANGUAGE plpgsql VOLATILE
AS
$$
  DECLARE
    -- the ddl we're building
    v_table_ddl text;

    -- data about the target table
    v_table_oid int;

    -- records for looping
    v_column_record record;
    v_constraint_record record;
    v_index_record record;
  BEGIN
    -- grab the oid of the table; https://www.postgresql.org/docs/8.3/catalog-pg-class.html
    SELECT c.oid INTO v_table_oid
    FROM pg_catalog.pg_class c
    LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
    WHERE 1=1
      AND c.relkind = 'r' -- r = ordinary table; https://www.postgresql.org/docs/9.3/catalog-pg-class.html
      AND c.relname = in_table_name -- the table name
      AND n.nspname = in_schema_name; -- the schema

    -- throw an error if table was not found
    IF (v_table_oid IS NULL) THEN
      RAISE EXCEPTION 'table does not exist';
    END IF;

    -- start the create definition
    v_table_ddl := 'CREATE TABLE ' || in_schema_name || '.' || in_table_name || ' (' || E'\n';

    -- define all of the columns in the table; https://stackoverflow.com/a/8153081/3068233
    FOR v_column_record IN
      SELECT
        c.column_name,
        c.data_type,
        COALESCE(
          c.character_maximum_length,
          CASE WHEN c.data_type = 'numeric' THEN c.numeric_precision ELSE null END -- only consider numeric_precision for the 'numeric' column, all others dont specify
        ) as precision,
        CASE WHEN c.numeric_scale IS NOT null AND c.numeric_scale <> 0 THEN c.numeric_scale ELSE null END as scale,
        c.is_nullable,
        c.column_default
      FROM information_schema.columns c
      WHERE (table_schema, table_name) = (in_schema_name, in_table_name)
      ORDER BY ordinal_position
    LOOP
      v_table_ddl := v_table_ddl || '  ' -- note: two char spacer to start, to indent the column
        || v_column_record.column_name || ' '
        || v_column_record.data_type
        || CASE WHEN v_column_record.precision IS NOT NULL THEN
          (
            '('
            || v_column_record.precision
            || CASE WHEN v_column_record.scale IS NOT NULL THEN (', ' || v_column_record.scale) ELSE '' END
            || ')'
          ) ELSE '' END || ' '
        || CASE WHEN v_column_record.is_nullable = 'NO' THEN 'NOT NULL' ELSE 'NULL' END
        || CASE WHEN v_column_record.column_default IS NOT null THEN (' DEFAULT ' || v_column_record.column_default) ELSE '' END
        || ',' || E'\n';
    END LOOP;

    -- define all the constraints in the; https://www.postgresql.org/docs/9.1/catalog-pg-constraint.html && https://dba.stackexchange.com/a/214877/75296
    FOR v_constraint_record IN
      SELECT
        con.conname as constraint_name,
        con.contype as constraint_type,
        CASE
          WHEN con.contype = 'p' THEN 1 -- primary key constraint
          WHEN con.contype = 'u' THEN 2 -- unique constraint
          WHEN con.contype = 'f' THEN 3 -- foreign key constraint
          WHEN con.contype = 'c' THEN 4
          ELSE 5
        END as type_rank,
        pg_get_constraintdef(con.oid) as constraint_definition
      FROM pg_catalog.pg_constraint con
      JOIN pg_catalog.pg_class rel ON rel.oid = con.conrelid
      JOIN pg_catalog.pg_namespace nsp ON nsp.oid = connamespace
      WHERE nsp.nspname = in_schema_name
      AND rel.relname = in_table_name
      ORDER BY type_rank, constraint_name -- pk > uuid > fk; fk0 > fk1
    LOOP
      v_table_ddl := v_table_ddl || '  ' -- note: two char spacer to start, to indent the column
        || 'CONSTRAINT' || ' '
        || v_constraint_record.constraint_name || ' '
        || v_constraint_record.constraint_definition
        || ',' || E'\n';
    END LOOP;

    -- drop the last comma before ending the create statement
    v_table_ddl = substr(v_table_ddl, 0, length(v_table_ddl) - 1) || E'\n';

    -- end the create definition
    v_table_ddl := v_table_ddl || ');' || E'\n';

    -- suffix create statement with all of the indexes on the table
    FOR v_index_record IN
      SELECT indexdef
      FROM pg_indexes i
      LEFT JOIN pg_catalog.pg_constraint c ON i.indexname = c.conname
      WHERE 1=1
        AND (schemaname, tablename) = (in_schema_name, in_table_name)
        AND c.contype is null -- and is not an index auto created for a constraint
      ORDER BY indexdef asc -- makes it so that "fk0" ordered before "fk1"
    LOOP
      v_table_ddl := v_table_ddl
        || v_index_record.indexdef
        || ';' || E'\n';
    END LOOP;

    -- return the ddl
    RETURN v_table_ddl;
  END;
$$;
