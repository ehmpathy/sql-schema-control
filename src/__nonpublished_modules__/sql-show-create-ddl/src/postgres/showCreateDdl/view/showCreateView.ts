import { DatabaseConnection } from '../../../types';

export const showCreateView = async ({
  dbConnection,
  schema,
  name,
}: {
  dbConnection: DatabaseConnection;
  schema: string;
  name: string;
}) => {
  const result = await dbConnection.query({
    sql: `
SELECT 'CREATE OR REPLACE VIEW ${name} AS ' || E'\n' || ' ' || pg_get_viewdef(c.oid) as ddl
FROM pg_catalog.pg_class c
LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
WHERE 1=1
  AND c.relkind = 'v' -- v = view; https://www.postgresql.org/docs/9.3/catalog-pg-class.html
  AND n.nspname = $1 -- the schema
  AND c.relname = $2; -- the view name
    `.trim(),
    values: [schema, name],
  });
  if (!result.rows) throw new Error(`could not find view '${schema}.${name}'`);
  return result.rows[0].ddl;
};
