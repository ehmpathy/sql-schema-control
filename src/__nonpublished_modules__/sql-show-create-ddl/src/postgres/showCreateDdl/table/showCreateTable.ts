import { DatabaseConnection } from '../../../types';
import { provisionShowCreateTableFunction } from './provisionShowCreateTableFunction';

export const showCreateTable = async ({
  dbConnection,
  schema,
  table,
}: {
  dbConnection: DatabaseConnection;
  schema: string;
  table: string;
}) => {
  // provision the show create function if it dne (since nothing like this comes standard w/ postgres)
  const { rows } = await dbConnection.query({
    sql: "select * from information_schema.routines where routine_name = 'show_create_table' ",
  });
  if (!rows.length) {
    // since dne, provision it
    await provisionShowCreateTableFunction({ dbConnection });
  }

  // grab the ddl our function generates
  const result = await dbConnection.query({
    sql: 'SELECT public.show_create_table($1, $2) as ddl',
    values: [schema, table],
  });
  if (!result.rows) throw new Error(`could not find table '${schema}.${table}'`);
  return result.rows[0].ddl;
};
