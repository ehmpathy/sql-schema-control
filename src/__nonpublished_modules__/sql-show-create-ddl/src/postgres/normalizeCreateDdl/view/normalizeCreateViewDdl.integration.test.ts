import { DatabaseLanguage } from '../../../../../../domain';
import { getDbConnection } from '../../../__test_assets__/getDbConnection';
import { DatabaseConnection } from '../../../types';
import { provisionShowCreateTableFunction } from '../../showCreateDdl/table/provisionShowCreateTableFunction';
import { showCreateView } from '../../showCreateDdl/view/showCreateView';
import { provisionDependencyTablesForMoreComplexView } from './__test_assets__/provisionDependencyTablesForMoreComplexView';
import { normalizeCreateViewDdl } from './normalizeCreateViewDdl';

describe('showCreateView', () => {
  let dbConnection: DatabaseConnection;
  beforeAll(async () => {
    dbConnection = await getDbConnection({
      language: DatabaseLanguage.POSTGRES,
    });
  });
  afterAll(async () => {
    await dbConnection.end();
  });
  it('should be possible to normalize create view ddl', async () => {
    await provisionShowCreateTableFunction({ dbConnection });
    await dbConnection.query({
      sql: 'DROP VIEW IF EXISTS test_view_for_normalization_on;',
    });
    await dbConnection.query({
      sql: "CREATE VIEW test_view_for_normalization_on as SELECT 'hello' as first_words",
    });
    const ddl = await showCreateView({
      dbConnection,
      schema: 'public',
      name: 'test_view_for_normalization_on',
    });
    const normalizedDdl = await normalizeCreateViewDdl({ ddl });
    expect(normalizedDdl).toMatchSnapshot();
  });
  it('should normalize a more complex create view ddl cleanly', async () => {
    await dbConnection.query({ sql: 'DROP VIEW IF EXISTS view_home_current;' });
    await provisionDependencyTablesForMoreComplexView({ dbConnection });
    const complexViewDdl = `
CREATE OR REPLACE VIEW view_home_current AS
SELECT
  s.id,
  s.uuid,
  s.name,
  (
    SELECT array_agg(home_to_host.host_id ORDER BY home_to_host.array_order_index) as array_agg
    FROM home_to_host WHERE home_to_host.home_id = s.id
  ) as host_ids,
  s.built,
  s.bedrooms,
  s.bathrooms,
  (
    SELECT array_agg(home_version_to_photo.photo_id ORDER BY home_version_to_photo.array_order_index) as array_agg
    FROM home_version_to_photo WHERE home_version_to_photo.home_version_id = v.id
  ) as photo_ids,
  s.created_at,
  v.effective_at,
  v.created_at as updated_at
FROM home s
JOIN home_cvp cvp ON s.id = cvp.home_id
JOIN home_version v ON v.id = cvp.home_version_id;
    `.trim();
    await dbConnection.query({
      sql: complexViewDdl,
    });
    const ddl = await showCreateView({
      dbConnection,
      schema: 'public',
      name: 'view_home_current',
    });
    const normalizedDdl = await normalizeCreateViewDdl({ ddl });
    expect(normalizedDdl).toMatchSnapshot();

    // and it should normalize a reasonable definition of the view and the show create in the same way, since they are the same fundamentally
    const normalizedSourceDdl = normalizeCreateViewDdl({ ddl: complexViewDdl });
    expect(normalizedDdl).toEqual(normalizedSourceDdl); // should be equivalent, since they both describe the same view - and the source ddl is a reasonable expression of it
  });
});
