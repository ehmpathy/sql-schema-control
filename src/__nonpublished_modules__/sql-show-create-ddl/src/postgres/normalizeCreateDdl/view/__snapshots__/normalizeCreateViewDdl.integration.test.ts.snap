// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`showCreateView should be possible to normalize create view ddl 1`] = `
"CREATE OR REPLACE VIEW test_view_for_normalization_on AS
select
  'hello' as first_words"
`;

exports[`showCreateView should normalize a more complex create view ddl cleanly 1`] = `
"CREATE OR REPLACE VIEW view_home_current AS
select
  s.id,
  s.uuid,
  s.name,
  (
    select
      array_agg(
        home_to_host.host_id
        order by
          home_to_host.array_order_index
      ) as array_agg
    from
      home_to_host
    where
      home_to_host.home_id = s.id
  ) as host_ids,
  s.built,
  s.bedrooms,
  s.bathrooms,
  (
    select
      array_agg(
        home_version_to_photo.photo_id
        order by
          home_version_to_photo.array_order_index
      ) as array_agg
    from
      home_version_to_photo
    where
      home_version_to_photo.home_version_id = v.id
  ) as photo_ids,
  s.created_at,
  v.effective_at,
  v.created_at as updated_at
from
  home s
  join home_cvp cvp on s.id = cvp.home_id
  join home_version v on v.id = cvp.home_version_id"
`;
